x/ibc-rate-limit/contracts/rate-limiter/src/state/flow.rs

```diff
@@ -0,0 +1,167 @@
+/// A Flow represents the transfer of value for a denom through an IBC channel
+/// during a time window.
+///
+/// It tracks inflows (transfers into osmosis) and outflows (transfers out of
+/// osmosis).
+///
+/// The period_end represents the last point in time for which this Flow is
+/// tracking the value transfer.
+///
+/// Periods are discrete repeating windows. A period only starts when a contract
+/// call to update the Flow (SendPacket/RecvPackt) is made, and not right after
+/// the period ends. This means that if no calls happen after a period expires,
+/// the next period will begin at the time of the next call and be valid for the
+/// specified duration for the quota.
+///
+/// This is a design decision to avoid the period calculations and thus reduce gas consumption
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema, Copy)]
+pub struct Flow {
+    pub inflow: Uint256,
+    pub outflow: Uint256,
+    pub period_end: Timestamp,
+}
+
+impl Flow {
+    pub fn new(
+        inflow: impl Into<Uint256>,
+        outflow: impl Into<Uint256>,
+        now: Timestamp,
+        duration: u64,
+    ) -> Self {
+        Self {
+            inflow: inflow.into(),
+            outflow: outflow.into(),
+            period_end: now.plus_seconds(duration),
+        }
+    }
+
+    /// The balance of a flow is how much absolute value for the denom has moved
+    /// through the channel before period_end. It returns a tuple of
+    /// (balance_in, balance_out) where balance_in in is how much has been
+    /// transferred into the flow, and balance_out is how much value transferred
+    /// out.
+    pub fn balance(&self) -> (Uint256, Uint256) {
+        (
+            self.inflow.saturating_sub(self.outflow),
+            self.outflow.saturating_sub(self.inflow),
+        )
+    }
+
+    /// checks if the flow, in the current state, has exceeded a max allowance
+    pub fn exceeds(&self, direction: &FlowType, max_inflow: Uint256, max_outflow: Uint256) -> bool {
+        let (balance_in, balance_out) = self.balance();
+        match direction {
+            FlowType::In => balance_in > max_inflow,
+            FlowType::Out => balance_out > max_outflow,
+        }
+    }
+
+    /// returns the balance in a direction. This is used for displaying cleaner errors
+    pub fn balance_on(&self, direction: &FlowType) -> Uint256 {
+        let (balance_in, balance_out) = self.balance();
+        match direction {
+            FlowType::In => balance_in,
+            FlowType::Out => balance_out,
+        }
+    }
+
+    /// If now is greater than the period_end, the Flow is considered expired.
+    pub fn is_expired(&self, now: Timestamp) -> bool {
+        self.period_end < now
+    }
+
+    // Mutating methods
+
+    /// Expire resets the Flow to start tracking the value transfer from the
+    /// moment this method is called.
+    pub fn expire(&mut self, now: Timestamp, duration: u64) {
+        self.inflow = Uint256::from(0_u32);
```

Comment:
Nit: Super small but using `Uint256::zero()` is a bit cleaner. Same for line below.

Reply:
```suggestion
        self.inflow = Uint256::zero()
```
Sure! sounds good!